velox/expression/tests/ExpressionFuzzer.cpp

```diff
@@ -513,36 +526,52 @@ std::vector<core::TypedExprPtr> ExpressionFuzzer::generateRegexpReplaceArgs(
   return inputExpressions;
 }
 
+// Either generates a new expression of the required return type or if already
+// generated expressions of the same return type exist then there is a 30%
+// chance that it will re-use one of them. Only expressions with no nested
+// expressions are re-used.
 core::TypedExprPtr ExpressionFuzzer::generateExpression(
     const TypePtr& returnType) {
   VELOX_CHECK_GT(remainingLevelOfNesting_, 0);
   --remainingLevelOfNesting_;
 
   auto guard = folly::makeGuard([&] { ++remainingLevelOfNesting_; });
 
-  auto firstAttempt =
-      &ExpressionFuzzer::generateExpressionFromConcreteSignatures;
-  auto secondAttempt =
-      &ExpressionFuzzer::generateExpressionFromSignatureTemplate;
-
-  size_t useSignatureTemplate =
-      boost::random::uniform_int_distribution<uint32_t>(0, 1)(rng_);
-  if (FLAGS_velox_fuzzer_enable_complex_types && useSignatureTemplate) {
-    std::swap(firstAttempt, secondAttempt);
-  }
-
-  core::TypedExprPtr expression = (this->*firstAttempt)(returnType);
-  if (!expression) {
-    if (FLAGS_velox_fuzzer_enable_complex_types) {
-      expression = (this->*secondAttempt)(returnType);
+  auto& listOfCandidateExprs = typeToExpressions_[returnType->toString()];
+  bool addNewExpr = remainingLevelOfNesting_ > 0 ||
```

Comment:
The branch that calls generateExpressionFromConcreteSignatures and generateExpressionFromSignatureTemplate does not need this check. However, the branch that reuse an existing expression does need to check that the current level of nesting plus the level of the reused expression doesn't exceed the threshold.

Reply:
Right, will remove this.
This check was initially added to ensure that we only re-use if we are at the end of the nesting level because we only store single level expressions. Removing this can re-use expressions at higher remainingLevelOfNesting_ which is also fine since we can also end up with randomly generated expressions that end earlier even when we have more levels left